src/h3lib/lib/h3Index.c

```diff
@@ -119,50 +119,183 @@ H3Error H3_EXPORT(h3ToString)(H3Index h, char *str, size_t sz) {
+| 110.111 | 100.100 | 001.000 | 111.111* |              100.100 | invalid |
+| 110.110 | 100.100 | 001.001 | 000.000  |              000.000 | OK      |
+
+  *: carry happened
+
+In the second example with 110.111, we "misidentify" the 110 as a 7, due
+to a carry from the lower bits. But this is OK because we correctly
+identify the lowest (only, in this example) 7 just before it.
 
-        if (NEVER(digit < CENTER_DIGIT) || digit >= NUM_DIGITS) return 0;
+We only have to worry about carries affecting higher bits in the case of
+a 7; all other digits (0--6) don't cause a carry when computing ~d - MLO.
+So even though a 7 can affect the results of higher bits, this is OK
+because we will always correctly identify the lowest 7.
+
+For further notes, see the discussion here:
+https://github.com/uber/h3/pull/496#discussion_r795851046
+*/
+static inline bool _hasAny7UptoRes(H3Index h, int res) {
+    const uint64_t MHI = 0b100100100100100100100100100100100100100100100;
+    const uint64_t MLO = MHI >> 2;
+
+    int shift = 3 * (15 - res);
+    h >>= shift;
+    h <<= shift;
+    h = (h & MHI & (~h - MLO));
+
+    return h != 0;
+}
+
+/* Check that all unused digits after `res` are set to 7 (INVALID_DIGIT).
+
+Bit shift to avoid looping through digits.
+*/
+static inline bool _hasAll7AfterRes(H3Index h, int res) {
+    // NOTE: res check is needed because we can't shift by 64
+    if (res < 15) {
+        int shift = 19 + 3 * res;
+
+        h = ~h;
+        h <<= shift;
+        h >>= shift;
+
+        return h == 0;
     }
+    return true;
+}
+
+/*
+Get index of first nonzero bit of an H3Index.
+
+When available, use compiler intrinsics, which should be fast.
+If not available, fall back to a loop.
+*/
+static inline int _firstOneIndex(H3Index h) {
+#if defined(__GNUC__) || defined(__clang__)
+    return 63 - __builtin_clzll(h);
+#elif defined(_MSC_VER) && defined(_M_X64)  // doesn't work on win32
+    unsigned long index;
+    _BitScanReverse64(&index, h);
+    return (int)index;
+#else
+    // Portable fallback
+    int pos = 63 - 19;
+    H3Index m = 1;
+    while ((h & (m << pos)) == 0) pos--;
+    return pos;
+#endif
+}
 
-    for (int r = res + 1; r <= MAX_H3_RES; r++) {
-        Direction digit = H3_GET_INDEX_DIGIT(h, r);
-        if (digit != INVALID_DIGIT) return 0;
+/*
+One final validation just for cells whose base cell (res 0)
+is a pentagon.
+
+Pentagon cells start with a sequence of 0's (CENTER_DIGIT's).
+The first nonzero digit can't be a 1 (i.e., "deleted subsequence",
+PENTAGON_SKIPPED_DIGIT, or K_AXES_DIGIT).
+
+We can check that (in the lower 45 = 15*3 bits) the position of the
+first 1 bit isn't divisible by 3.
+*/
+static inline bool _hasDeletedSubsequence(H3Index h, int base_cell) {
+    static const bool isBaseCellPentagonArr[128] = {
+        [4] = 1,  [14] = 1, [24] = 1, [38] = 1, [49] = 1,  [58] = 1,
+        [63] = 1, [72] = 1, [83] = 1, [97] = 1, [107] = 1, [117] = 1};
+
+    if (isBaseCellPentagonArr[base_cell]) {
```

Comment:
Is there a reason not to use (perhaps an inlined version of) `_isBaseCellPentagon`?

Reply:
Good point. I'll run some benchmarks and see if there's any difference.

Reply:
Definitely slower with the naive use of `_isBaseCellPentagon`. Test setup in the latest commit:

<img width="635" alt="image" src="https://github.com/user-attachments/assets/150ffb62-1494-444a-9e5b-9c96e14ca62d" />

I'll figure out the right `extern` and `inline` incantation and try that next. 

Another alternative might be to just expose the `BaseCellData` directly.

More generally for the H3 lib as a whole, I'm not sure if there's a benefit to using a struct of arrays vs an array of structs.

Reply:
I think doing this will be a little involved, so I'm proposing doing it later :)

https://github.com/uber/h3/issues/984